### ISSUE TITLE
Meta tag_key Inconsistent Behavior (>=29.2)

### ISSUE THREAD

**Issue by Steven-Hogue** [rnag/dataclass-wizard#148]
* Dataclass Wizard version: v0.29.1 / v0.29.2-3
* Python version: 3.11.2
* Operating System: Windows/Linux

### Description

I have historically had a system that had one base type, with many alternatives for a field depending on the sub types. For example, a "Result" type with a `data` field that could be an "XML" or "HTML" type (purely for example).

However, on `dataclass-wizard>=0.29.2`, all Unix-like systems I have tested on (Git Bash, Debian 12, RHEL 9) do not permit this. It seems like it may somehow be consuming the tag field and not allowing it to be used as a part of the class structure. Oddly though, I have had no issues on Windows systems.

### What I Did

Here is the example code (again, not exact use-case):

```python
from dataclasses import dataclass

from dataclass_wizard import JSONWizard

@dataclass
class DataType(JSONWizard):
    id: int
    type: str

@dataclass
class XML(DataType):
    class _(JSONWizard.Meta):
        tag = "xml"
    field_type_1: str

@dataclass
class HTML(DataType):
    class _(JSONWizard.Meta):
        tag = "html"
    field_type_2: str

@dataclass
class Result(JSONWizard):
    class _(JSONWizard.Meta):
        tag_key = "type"
    data: XML | HTML
```

Here is the result on Windows, and Linux prior to dataclass-wizard==0.29.2:
```python
>>> t1 = Result.from_dict(
...     {"data": {"id": 1, "type": "xml", "field_type_1": "value"}}
... )
>>> t2 = Result.from_dict(
...     {"data": {"id": 2, "type": "html", "field_type_2": "value"}}
... )

>>> t1, t2
(Result(data=Type1(id=1, type='xml', field_type_1='value')),
Result(data=Type2(id=2, type='html', field_type_2='value')))
```

However, on Linux with dataclass-wizard>=0.29.2, this is the result:
```python
>>> t1 = Result.from_dict(
...     {"data": {"id": 1, "type": "xml", "field_type_1": "value"}}
... )
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "######/python3.11/site-packages/dataclass_wizard/loaders.py", line 571, in fromdict
    return load(d)
           ^^^^^^^
  File "<string>", line 21, in cls_fromdict
  File "######/python3.11/site-packages/dataclass_wizard/parsers.py", line 296, in __call__
    return self.tag_to_parser[tag](o)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<string>", line 35, in cls_fromdict
dataclass_wizard.errors.MissingFields: Failure calling constructor method of class `XML`. Missing values for required dataclass fields.
  have fields: ['id', 'field_type_1']
  missing fields: ['type']
  input JSON object: {"id": 1, "type": "xml", "field_type_1": "value"}
  error: XML.__init__() missing 1 required positional argument: 'type'
  key transform: to_snake_case()
  resolution: For more details, please see https://github.com/rnag/dataclass-wizard/issues/54
>>> 
>>> t2 = Result.from_dict(
...     {"data": {"id": 2, "type": "html", "field_type_2": "value"}}
... )
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<string>", line 21, in cls_fromdict
  File "######/python3.11/site-packages/dataclass_wizard/parsers.py", line 296, in __call__
    return self.tag_to_parser[tag](o)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<string>", line 35, in cls_fromdict
dataclass_wizard.errors.MissingFields: Failure calling constructor method of class `HTML`. Missing values for required dataclass fields.
  have fields: ['id', 'field_type_2']
  missing fields: ['type']
  input JSON object: {"id": 2, "type": "html", "field_type_2": "value"}
  error: HTML.__init__() missing 1 required positional argument: 'type'
  key transform: to_snake_case()
  resolution: For more details, please see https://github.com/rnag/dataclass-wizard/issues/54
```

Other than downgrading to version `dataclass-wizard==0.29.1`, another potential fix is to modify the `DataType` class to not accept the tag field as follows:
```python
@dataclass
class DataType(JSONWizard):
    id: int
#    type: str
```
And this functions fine unless you need to use that type for further analysis, which in my use-case I do.

**Comment by Steven-Hogue** [comment]
In further testing, commenting out the following lines seems to resolve the issue, so something about their implementation adversely effects normal tagging:

https://github.com/rnag/dataclass-wizard/blob/2fcf4cb850fec740f116c574e9e5d5ae3a910525/dataclass_wizard/loaders.py#L645C1-L647C51

**Comment by rnag** [comment]
Hi @Steven-Hogue , thanks for opening this issue. I agree this is certainly puzzling behavior, and I'll have to spend time to look into what the issue is 🤔 .

While I was able to confirm that you are correct that manually setting tags for each class results in the error, for some reason setting `auto_assign_tags = True` seems to fix the issue. This is bizarre, and I'll need to spend time to look into it to figure out what the difference b/w the two versions is.
 
Here is my working code using `auto_assign_tags`:

```python3
from dataclasses import dataclass

from dataclass_wizard import JSONWizard

@dataclass
class DataType(JSONWizard):
    id: int
    type: str

@dataclass
class XML(DataType):
    field_type_1: str

@dataclass
class HTML(DataType):
    field_type_2: str

@dataclass
class Result(JSONWizard):
    class _(JSONWizard.Meta):
        tag_key = "type"
        auto_assign_tags = True
    data: XML | HTML

t1 = Result.from_dict(
    {"data": {"id": 1, "type": "XML", "field_type_1": "value"}}
)

t2 = Result.from_dict(
{"data": {"id": 2, "type": "HTML", "field_type_2": "value"}}
)

from pprint import pprint
pprint((t1, t2))
```

Outputs:

```
(Result(data=XML(id=1, type='XML', field_type_1='value')),
 Result(data=HTML(id=2, type='HTML', field_type_2='value')))
```

**Comment by rnag** [comment]
Follow-up again (sorry for multiple posts), but I noticed that it works if I comment out the `type` field in `DataType`:

```python3
from dataclasses import dataclass

from dataclass_wizard import JSONWizard


@dataclass
class DataType(JSONWizard):
    id: int
    # If I uncomment below, the error occurs.
    # type: str

@dataclass
class XML(DataType):
    class _(JSONWizard.Meta):
        tag = "xml"
    field_type_1: str

@dataclass
class HTML(DataType):
    class _(JSONWizard.Meta):
        tag = "html"
    field_type_2: str

@dataclass
class Result(JSONWizard):
    class _(JSONWizard.Meta):
        tag_key = "type"
    data: XML | HTML

t1 = Result.from_dict(
    {"data": {"id": 1, "type": "xml", "field_type_1": "value"}}
)

t2 = Result.from_dict(
{"data": {"id": 2, "type": "html", "field_type_2": "value"}}
)

from pprint import pprint
pprint((t1, t2))
```

I think that's because the `tag_key = "type"` is just `Union`-specific data, it wouldn't normally be expected to be defined as a field on any datacalsses in the `Union`.

However, that poses a problem if the dataclass defines the tag key as a dataclass field 🤔  -- and also, I'm not sure about the inconsistent behavior between defining tags and `auto_assign_tags=True`? I'll certainly need to carve out some time to look into this more in detail, but for now I hope my findings are able to shed some light on the issue.

**Comment by rnag** [comment]
Kind of a "workaround" for now, omitting `type` field as you noted, but in case you need the `type` for further analysis:

```python3
from dataclasses import dataclass

from dataclass_wizard import JSONWizard
from dataclass_wizard.decorators import cached_class_property


@dataclass
class DataType(JSONWizard):
    id: int
    # If I uncomment below, the error occurs.
    # type: str

    @cached_class_property
    def type(cls):
        return cls.__name__.lower()
        # below is a kind of hacky solution that also works.
        # return getattr(cls._, 'tag', None)

@dataclass
class XML(DataType):
    class _(JSONWizard.Meta):
        tag = "xml"
    field_type_1: str

@dataclass
class HTML(DataType):
    class _(JSONWizard.Meta):
        tag = "html"
    field_type_2: str

@dataclass
class Result(JSONWizard):
    class _(JSONWizard.Meta):
        tag_key = "type"
    data: XML | HTML

t1 = Result.from_dict(
    {"data": {"id": 1, "type": "xml", "field_type_1": "value"}}
)

t2 = Result.from_dict(
{"data": {"id": 2, "type": "html", "field_type_2": "value"}}
)

from pprint import pprint
objects = t1, t2
pprint(objects)
print()
for o in objects:
    print(o.data.type)
```

**Comment by rnag** [comment]
@Steven-Hogue this should be fixed as of `v0.30.1`! Also added a [test case](https://github.com/rnag/dataclass-wizard/pull/150/files#diff-e05972075379c18a436be8fbf94140a60ff36d2758167f4421ff743a8aee4107R2444-R2477) to confirm intended behavior.

> In further testing, commenting out the following lines seems to resolve the issue, so something about their implementation adversely effects normal tagging on _some_ systems:
> 
> https://github.com/rnag/dataclass-wizard/blob/2fcf4cb850fec740f116c574e9e5d5ae3a910525/dataclass_wizard/loaders.py#L645C1-L647C51

Good eye, you spotted it quicker than me! After puzzling over it for a bit,  realized that the resolution was actually simple. If we set an `ExplicitNull` for the `tag_key`, that means it's skipped and not passed to `__init__()` at all. So I just added a check that it's not defined as a dataclass field. Appears to be fixed now! 🎉
